File: c/SRC/PLAYER/player.c
#include "player.h"

t_player	*init_player(double pos_x, double pos_y,
							double dir_x, double dir_y)
{
	t_player	*player;

	if (!(player = malloc(sizeof(t_player))))
		return (NULL);
	player->pos[X] = pos_x;
	player->pos[Y] = pos_y;
	player->dir[X] = dir_x;
	player->dir[Y] = dir_y;
	return (player);
}

void		behavior(t_player *player, int key)
{
	if (key == FORWARD)
		move(player->dir, player->pos, MOVE_SPEED);
	if (key == BACK)
		move(player->dir, player->pos, -MOVE_SPEED);
	if (key == LEFT)
		rotate(player->dir, ROT_SPEED);
	if (key == RIGHT)
		rotate(player->dir, -ROT_SPEED);
}

File: js/SRC/PLAYER/player.js
const X = 0;
const Y = 1;

class Player {
	constructor (info) {
		this.pos = info.initPos;
		this.dir = info.initDir;
		this.moveSpeed = info.moveSpeed;
		this.rotSpeed = info.rotSpeed;
	}

	move(speed) {
		this.pos = this.pos.map((e, i) =>
			e += this.dir[i] * speed);
	}

	rotate(speed) {
		const angle = speed * Math.PI / 180;
		
		//TODO : have to chcek!
		this.dir = [this.dir[X] * Math.cos(angle)
						- this.dir[Y] * Math.sin(angle),
					this.dir[X] * Math.sin(angle)
						+ this.dir[Y] * Math.sin(angle)];
	}

	behavior(key) {
		if (key === "FORWARD")
			this.move(this.moveSpeed);
		if (key === "BACK")
			this.move(-this.moveSpeed);
		if (key === "LEFT")
			this.rotate(this.rotSpeed);
		if (key === "RIGHT")
			this.rotate(-this.rotSpeed);
	}
}

module.exports = Player;